File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys
input = sys.stdin.readline
from collections import deque

V = int(input())
A = [[] for _ in range(V + 1)]
visited = [False] * (V + 1)
distance = [0] * (V + 1)

for i in range(V):
    numbers = list(map(int, input().split()))
    start = numbers[0]
    cnt = 1
    while numbers[cnt] != -1:
        A[start].append((numbers[cnt], numbers[cnt + 1]))
        cnt += 2

# print(A)
def BFS(v):
    queue = deque()
    queue.append(v)
    visited[v] = True

    while queue:
        now = queue.popleft()

        for i in A[now]:
            if not visited[i[0]]:
                queue.append(i[0])
                visited[i[0]] = True
                if distance[i[0]] < distance[now] + i[1]:
                    distance[i[0]] = distance[now] + i[1]
BFS(1)
# print(distance)
# print(visited)
Max = 1

for i in range(2, V + 1):
    if distance[i] > distance[Max]:
        Max = i
# print(Max)
visited = [False] * (V + 1)
distance = [0] * (V + 1)
BFS(Max)
print(max(distance))

File: 백준/Gold/1967. 트리의 지름/트리의 지름.cc
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <set>
#include <map>
#include <queue>
#include <deque>
#include <stack>
#include <cstring>

using namespace std;
#define fastio ios::sync_with_stdio(false); cin.tie(0); cout.tie(nullptr);

struct Edge{
    int node, length;
};

const int MX = 10001;

int N;
vector<Edge> adj[MX];
int visited[MX];
int result;

int bfs(int start){
    memset(visited, -1, sizeof(visited));
    queue<int> q;
    q.push(start);
    visited[start] = 0;

    int maxNode=0;

    while(!q.empty()){
        int now = q.front(); q.pop();

        if(visited[now] > visited[maxNode]){
            maxNode = now;
        }

        for(Edge edge : adj[now]){
            if(visited[edge.node] != -1) continue;

            q.push(edge.node);
            visited[edge.node] = visited[now] + edge.length;
        }
    }

    return maxNode;
}


int main(void){
    fastio;
    
    // freopen("input.txt", "r", stdin);

    cin >> N;
    int a, b, c;
    for(int i=0; i<N-1; i++){
        cin >> a >> b >> c;
        adj[a].push_back({b, c});
        adj[b].push_back({a, c});
    }

    int left = bfs(1);
    int right = bfs(left);

    cout << visited[right] << '\n';

    return 0;
}